lib/pbench/server/cache_manager.py

```diff
@@ -861,16 +885,26 @@ def _discover_tarballs(self):
         """Discover the known tarballs
 
         Look in the ARCHIVE tree's controller directory for tarballs, and add
-        them to the known set. We also check for unpacked directories in the
-        CACHE tree matching the resource_id of any tarballs we find in order
-        to link them.
+        them to the known set. "Old" tarballs may be at the top level, "new"
+        tarballs are in "resource_id" isolation directories.
+
+        We also check for unpacked directories in the CACHE tree matching the
+        resource_id of any tarballs we find in order to link them.
         """
         for file in self.path.iterdir():
             if file.is_file() and Dataset.is_tarball(file):
-                tarball = Tarball(file, self)
+                tarball = Tarball(file, get_tarball_md5(file), self)
                 self.tarballs[tarball.name] = tarball
                 self.datasets[tarball.resource_id] = tarball
                 tarball.check_unpacked()
+            elif file.is_dir():
+                md5 = file.name
+                for tar in file.iterdir():
+                    if tar.is_file() and Dataset.is_tarball(tar):
+                        tarball = Tarball(tar, md5, self)
+                        self.tarballs[tarball.name] = tarball
+                        self.datasets[tarball.resource_id] = tarball
+                        tarball.check_unpacked()
```

Comment:
This code also appears at lines 895-899...consider refactoring it into a helper method.

Reply:
I thought of that, but decided that it was either more complicated than it was worth (if I pulled the `if`) or so trivial as to be pointless if I didn't. Now, as often happens, you've got me teetering on the edge of the precipice ...

Reply:
It seemed like "enough" that it warranted refactoring.  Although, since both instances of the code blob fit on the same screen, there's an argument that the need for refactoring is diminished.  Still, it's not that hard to do (I think it's just three parameters plus `self`), and it would guard against future "drift".  (Since this is Python, you could even do it with an inner function, if that appeals to you.)

I think PyCharm will do the refactoring automagically (and it might actually apply it in both places -- at one point it was complaining at me about my duplicated code, so I would hope that it could fix it...), but you're using VSCode, right?